src/io/read.rs

```diff
@@ -0,0 +1,65 @@
+use std::mem;
+use std::io::Read;
+
+use futures::{Async, Future, Poll};
+
+enum State<R, T> {
+    Pending {
+        rd: R,
+        buf: T,
+    },
+    Empty,
+}
+
+fn eof() -> ::std::io::Error {
+    ::std::io::Error::new(::std::io::ErrorKind::UnexpectedEof, "unexpected EOF")
+}
+
+/// Baz.
+pub fn read_some<R, T>(rd: R, buf: T) -> ReadSome<R, T>
+    where R: Read,
+          T: AsMut<[u8]>
+{
+    ReadSome {
+        state: State::Pending {
+            rd: rd,
+            buf: buf,
+        }
+    }
+}
+
+/// Bar.
+pub struct ReadSome<R, T> {
+    state: State<R, T>,
+}
+
+impl<R, T> Future for ReadSome<R, T>
+    where R: Read,
+          T: AsMut<[u8]>
+{
+    type Item = (R, T, usize);
+    type Error = ::std::io::Error;
+
+    fn poll(&mut self) -> Poll<(R, T, usize), ::std::io::Error> {
+        let nread = match self.state {
+            State::Pending { ref mut rd, ref mut buf } => {
+                let buf = buf.as_mut();
+
+                match rd.read(&mut buf[..]) {
+                    Ok(0) => return Err(eof()),
```

Comment:
Hmm, if the buffer is zero-sized, then we just want to poll the socket. In this case reading of 0 bytes indicates about either some bytes available or FIN packet received. That's not an error.

On the other side, if the buffer is not zero-sized (which is likely), then reading 0 bytes is surely FIN.

I don't like an idea to make special case for zero-sized buffers. Instead I suggest to make this functionality as an async twin of `Read::read` method.